server/src/main/java/org/opensearch/grpc/services/nodesInfo/NodesInfoServiceImpl.java

```diff
@@ -0,0 +1,138 @@
+
+    public NodesInfoServiceImpl(NodeClient client) {
+        this.client = client;
+    }
+
+    @Override
+    public void getNodesInfo(NodesInfoProtoService.NodesInfoRequestProto request, StreamObserver<NodesInfoProtoService.NodesInfoResponseProto> responseObserver) {
+        NodesInfoResponse response = client.admin().cluster()
+            .nodesInfo(reqFromProto(request))
+            .actionGet();
+        responseObserver.onNext(respToProto(response));
+        responseObserver.onCompleted();
+    }
+    
+    private static NodesInfoRequest reqFromProto(NodesInfoProtoService.NodesInfoRequestProto request) {
+        String[] nodeIds = request.getNodeIdsList().toArray(new String[0]);
+        NodesInfoRequest nodesInfoRequest = new NodesInfoRequest(nodeIds);
+        nodesInfoRequest.timeout(request.getTimeout());
+
+        if (request.getMetrics().getAll()) {
+            nodesInfoRequest.all();
+        } else {
+            for (String met : request.getMetrics().getMetricList().getStrListList()) {
+                nodesInfoRequest.addMetric(met);
+            }
+        }
+
+        return nodesInfoRequest;
+    }
+
+    private static NodesInfoProtoService.NodesInfoResponseProto respToProto(NodesInfoResponse response) {
+        NodesInfoProtoService.NodesInfoResponseProto.Builder builder = NodesInfoProtoService.NodesInfoResponseProto.newBuilder();
+
+        for (NodeInfo ni : response.getNodes()) {
+            NodesInfoProto.NodesInfo.Builder nib = NodesInfoProto.NodesInfo.newBuilder()
+                .setNodeId(ni.getNode().getId())
+                .setNodeName(ni.getNode().getName())
+                .setTransport(ni.getNode().getAddress().toString())
+                .setHost(ni.getNode().getHostName())
+                .setIp(ni.getNode().getHostAddress())
+                .setIp(ni.getVersion().toString())
+                .setBuildType(ni.getBuild().type().displayName())
+                .setBuildType(ni.getBuild().hash());
+
+            if (ni.getTotalIndexingBuffer() != null) {
+                nib.setTotalIndexingBuffer(ni.getTotalIndexingBuffer().toString());
+            }
+
+            for (DiscoveryNodeRole role : ni.getNode().getRoles()) {
+                nib.addRoles(role.roleName());
+            }
+
+            for (Map.Entry<String, String> entry : ni.getNode().getAttributes().entrySet()) {
+                nib.putAttributes(entry.getKey(), entry.getValue());
+            }
+
+            try {
+                XContentBuilder settingsBuilder = MediaTypeRegistry.contentBuilder(MediaTypeRegistry.JSON);
+                String settingsXCont = ni.getSettings().toXContent(settingsBuilder, EMPTY_PARAMS).toString();
+                nib.setSettings(settingsXCont);
+            } catch (IOException e) {
+                logger.debug("Failed to build NodesInfo settings: " + e);
+            }
+
+            Optional.ofNullable(getNodeInfoJSON(ni, OsInfo.class)).ifPresent(nib::setOs);
```

Comment:
why we convert back to JSON? i assume the whole purpose is to get rid of xContent JSON parse logic and leverage the protobuf derser.

Reply:
NodesInfo removed